mypyc/genops.py

```diff
@@ -3395,15 +3399,91 @@ def visit_yield_expr(self, expr: YieldExpr) -> Value:
         next_block = BasicBlock()
         next_label = len(cls.blocks)
         cls.blocks.append(next_block)
-        self.assign(cls.next_label_target, self.add(LoadInt(next_label)), expr.line)
+        self.assign(cls.next_label_target, self.add(LoadInt(next_label)), line)
         self.add(Return(retval))
         self.activate_block(next_block)
 
-        self.add_raise_exception_blocks_to_generator_class(expr.line)
+        self.add_raise_exception_blocks_to_generator_class(line)
 
         assert cls.send_arg_reg is not None
         return cls.send_arg_reg
 
+    def visit_yield_from_expr(self, o: YieldFromExpr) -> Value:
+        # This is basically an implementation of the code in PEP 380.
+
+        # TODO: do we want to use the right types here?
+        result = self.alloc_temp(object_rprimitive)
+        to_yield_reg = self.alloc_temp(object_rprimitive)
+        received_reg = self.alloc_temp(object_rprimitive)
+        iter_reg = self.maybe_spill_assignable(
+            self.primitive_op(iter_op, [self.accept(o.expr)], o.line))
+
+        stop_block, main_block, done_block = BasicBlock(), BasicBlock(), BasicBlock()
+        _y_init = self.primitive_op(next_raw_op, [self.read(iter_reg)], o.line)
+        self.add(Branch(_y_init, stop_block, main_block, Branch.IS_ERROR))
+
+        # If the initial next() raised, assume it was a StopIteration and extract
+        # the value. If it wasn't, we keep raising.
+        self.activate_block(stop_block)
+        self.assign(result, self.primitive_op(check_stop_op, [], o.line), o.line)
+        self.goto(done_block)
+
+        self.activate_block(main_block)
+        self.assign(to_yield_reg, _y_init, o.line)
+
+        # OK Now the main loop!
+        loop_block = BasicBlock()
+        self.goto_and_activate(loop_block)
+
+        def try_body() -> None:
```

Comment:
I'd rather not use use nested functions. The surrounding method is too long in my opinion, which makes it hard to read and modify.

Reply:
I don't think I agree, in this case. I think the main problem here is that `yield from` is really involved, and it's necessarily going to be kind of complicated. But the nested function setup allows the structure to mimic the structure of the PEP 380 semantic definition (which should maybe be called out more explicitly). I think that pulling out the nested functions into methods and needing to pass in a bunch of arguments would confuse rather than clarify.

Reply:
I got rid of the nested functions in https://github.com/mypyc/mypyc/commit/b95eaaab7e18944d45caa2367b4e70a1a15e3161, and I think it does make things worse and not better.